runtime/signals.c

```diff
@@ -239,12 +245,16 @@ void caml_request_major_slice (void)
 {
   Caml_state->requested_major_slice = 1;
   caml_set_action_pending();
+  // update Caml_state->young_limit_c
+  caml_update_young_limit();
```

Comment:
I find this code (and the code below) rather fishy. If I read correctly:
- first we set `requested_major_slice = 1`
- then `set_action_pending` will set `something_to_do = 1` and call `notify_action`, which sets `young_limit = young_alloc_end`
- then `update_young_limit` will see that `requested_major_slice==1` and set `young_limit_c = young_alloc_end` and then (vacuously) `young_limit = young_limit_c`

There is a lot of redundancy in here, and the treatment of `young_limit_c` in `update_young_limit` feels rather ad-hoc.

I would propose the following as a simpler approach, but I am not sure that it is correct:
- Define a `caml_set_gc_pending` function as a simpler version of `caml_set_action_pending` for GC requests. This function sets `young_limit_c` and `young_limit` to `young_alloc_end` unconditionally.
- Use it here instead of `caml_set_action_pending` (no need to separate update the young limits).
- Remove the conditional in `caml_udpate_young_limit` from this patch (the new case is not used anymore, iiuc.) and only add an update to synchronize `young` and `young_c`.

Reply:
The test inside `caml_update_young_limit` intends to make the invariants of this functions less brittle, otherwise one has to be careful not to call it in places where a gc might be pending as it will forget the gc request. As a matter of fact, this seems to have been respected until recently, but currently if you call certain memprof functions after requesting a GC then the runtime could forget your request. Not a huge bug, but it shows that the invariants were too brittle.

Given that, it seemed like premature optimisation to me to introduce a specialised function to do the effect of update_young_limits without the redundancy. But since you seem to really dislike the detour, I tried to fix it differently.

Reply:
I'm not sure where you "tried to fix it differently", is it in the fixup commit https://github.com/ocaml/ocaml/pull/8961/commits/56af457dc60304e126735c16f9df4e4d4cb5da2d ? I'm slightly confused because this seems to be a simple commit factorizing two lines of code (I see no detour removal in that patch), but then it looks like at this point the detour is gone anyway (it contains code that I have not reviewed yet).

The detour is not a problem at runtime, but I find it confusing in terms of code comprehension: when we perform redundant operations it sometimes mean that the design could/should be sharpened.

Reply:
My rebase mistake. The fixup commit was intended to show the fix, but instead the fix is already in the original commit (set action_pending by hand without calling notify_action) and the fixup does a small refactoring.
The end version should indeed be less confusing for readers.